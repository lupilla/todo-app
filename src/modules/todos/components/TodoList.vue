<template>
  <main id="main" class="main" v-if="showCheckAllTodos">
    <div>
      <input
        id="checkAll"
        type="checkbox"
        class="toggle-all"
        data-id-test="checkAll"
        aria-label="Check all Todos"
        @click="changeTodosStatus"
        v-model="selected"
      />
      <label for="checkAll"></label>
    </div>
    <ul class="todo-list">
      <li
        v-for="todo in todos"
        :key="todo.id"
        data-id-test="edit"
        :class="[{ editing: todo.isEditing }, { completed: todo.completed }]"
        @dblclick="changeEditMode(todo)"
      >
        <TodoItem
          :todo="todo"
          @edit="editTodo"
          @change-status="changeTodoStatus(todo)"
          @stop-edit="stopEditing(todo)"
        ></TodoItem>
      </li>
    </ul>
  </main>
</template>
<script>
import TodoItem from "@/modules/todos/components/TodoItem.vue";
import { mapGetters, mapActions } from "vuex";

export default {
  name: "TodoList",
  data() {
    return {
      selected: false,
    };
  },
  components: {
    TodoItem,
  },
  props: {
    todos: {
      type: Array,
      default: () => [],
    },
  },
  methods: {
    ...mapActions("todos", [
      "updateTodo",
      "changeCompleteTodo",
      "changeCompletedTodos",
      "changeEditableTodo",
      "changeEditableTodos",
    ]),
    async changeEditMode(todo) {
      await this.changeEditableTodo(todo);
    },
    async changeTodosStatus() {
      this.selected = !this.selected;
      try {
        await this.changeCompletedTodos(this.selected);
      } catch (error) {
        console.error(error);
      }
    },
    async changeTodoStatus(todo) {
      try {
        await this.changeCompleteTodo(todo);
        this.statusChanged();
      } catch (error) {
        console.error(error);
      }
    },
    async stopEditing(todo) {
      await this.changeEditableTodo(todo);
    },
    async editTodo(todo) {
      try {
        await this.updateTodo(todo);
        await this.changeEditableTodos();
        this.statusChanged();
      } catch (error) {
        console.error(error);
      }
    },
    statusChanged() {
      if (this.$route.name === "allTodos" && this.getCompletedLength) {
        this.selected = true;
      } else if (this.$route.name === "activeTodos") {
        this.selected = false;
      } else if (this.$route.name === "completedTodos") {
        this.selected = true;
      } else {
        this.selected = false;
      }
    },
  },
  computed: {
    ...mapGetters("todos", ["completedTodos"]),
    showCheckAllTodos() {
      return this.todos.length > 0;
    },
    getCompletedLength() {
      return this.completedTodos.length === this.todos.length;
    },
  },
  watch: {
    "$route.name": {
      handler: function () {
        if (this.todos) {
          this.statusChanged();
        }
      },
      deep: true,
      immediate: true,
    },
    todos: {
      handler: function () {
        if (this.todos) {
          this.statusChanged();
        }
      },
      deep: true,
      immediate: true,
    },
  },
};
</script>
<style>
.main {
  position: relative;
  z-index: 2;
  border-top: 1px solid #e6e6e6;
}

.todo-list {
  margin: 0;
  padding: 0;
  list-style: none;
}

.todo-list li {
  position: relative;
  font-size: 24px;
  border-bottom: 1px solid #ededed;
}

.todo-list li:last-child {
  border-bottom: none;
}

.todo-list li.editing {
  border-bottom: none;
  padding: 0;
}

.todo-list li.editing .edit {
  display: block;
  width: 506px;
  padding: 12px 16px;
  margin: 0 0 0 43px;
}

.todo-list li.editing .view {
  display: none;
}

.todo-list li label {
  word-break: break-all;
  padding: 15px 15px 15px 60px;
  display: block;
  line-height: 1.2;
  transition: color 0.4s;
}

.todo-list li.editing:last-child {
  margin-bottom: -1px;
}

.todo-list li:hover .destroy {
  display: block;
}

.toggle-all {
  text-align: center;
  border: none; /* Mobile Safari */
  opacity: 0;
  position: absolute;
}

.toggle-all + label {
  width: 60px;
  height: 34px;
  font-size: 0;
  position: absolute;
  top: -52px;
  left: -13px;
  -webkit-transform: rotate(90deg);
  transform: rotate(90deg);
}

.toggle-all + label:before {
  content: "‚ùØ";
  font-size: 22px;
  color: #e6e6e6;
  padding: 10px 27px 10px 27px;
}

.toggle-all:checked + label:before {
  color: #737373;
}

/*
	Hack to remove background from Mobile Safari.
	Can't use it globally since it destroys checkboxes in Firefox
*/
@media screen and (-webkit-min-device-pixel-ratio: 0) {
  .toggle-all,
  .todo-list li .toggle {
    background: none;
  }

  .todo-list li .toggle {
    height: 40px;
  }
}
</style>
