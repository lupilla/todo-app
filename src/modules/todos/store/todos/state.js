/**
 * General state object of todo module
 * It contains basically an array with the todos collection
 */
export default () => ({
  todos: [],
});
