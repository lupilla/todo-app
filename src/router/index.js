import { createRouter, createWebHistory } from "vue-router";

/**
 * Array of routes included in the application
 * - Default route ("/"): all todos are showed
 * - Active route ("/active"): only active todos are showed
 * - Completed route ("/completed"): only completed todos are showed
 * - Page not found route: error page 404 is showed when the route does not match
 */
const routes = [
  {
    path: "/",
    name: "allTodos",
    component: () =>
      import(
        /* webpackChunkName: 'alltodos'*/ "@/modules/todos/pages/TodoPage.vue"
      ),
  },
  {
    path: "/active",
    name: "activeTodos",
    component: () =>
      import(
        /* webpackChunkName: 'active'*/ "@/modules/todos/pages/TodoPage.vue"
      ),
  },
  {
    path: "/completed",
    name: "completedTodos",
    component: () =>
      import(
        /* webpackChunkName: 'completed'*/ "@/modules/todos/pages/TodoPage.vue"
      ),
  },
  {
    path: "/:pathMatch(.*)*",
    component: () =>
      import(
        /* webpackChunkName: 'NoPageFoundPage'*/ "@/modules/shared/pages/NoPageFound.vue"
      ),
  },
];

/**
 * Router creation method to export it to the main.js file
 */
const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
});

export default router;
