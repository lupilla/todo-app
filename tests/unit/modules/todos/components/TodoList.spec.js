import { shallowMount } from "@vue/test-utils";
import TodoList from "@/modules/todos/components/TodoList.vue";
import { INITIAL_STATE } from "../../../mocks/state.js";
import createVuexStore from "../../../mocks/mock-store.js";
import router from "@/router";
const store = createVuexStore();
const $route = {
  path: "/",
};
describe("TodoList tests", () => {
  let wrapper;

  beforeEach(async () => {
    jest.clearAllMocks();
    wrapper = await shallowMount(TodoList, {
      global: {
        mocks: {
          $route,
        },
        plugins: [store, router],
      },
      props: {
        todos: INITIAL_STATE.todos,
      },
      computed: {
        showCheckAllTodos: () => true,
        getCompletedLength: () => false,
      },
    });
  });

  test("TodoList Component should match with snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("'selected' property should have a falsy value when all todos are not completed", () => {
    expect(wrapper.vm.$data.selected).toBe(false);
  });

  test("'changeTodosStatus' method should call 'changeCompletedTodos' action", async () => {
    const spyChangeTodosAction = jest.spyOn(wrapper.vm, "changeCompletedTodos");
    const checkAll = wrapper.find('[data-id-test="checkAll"]');
    await checkAll.trigger("click");
    expect(spyChangeTodosAction).toHaveBeenCalled();
  });

  test("'todos' property should have elements by default when todos are passed to the component", async () => {
    const todoItems = await wrapper.findAll("todo-item-stub");
    expect(wrapper.props("todos")).toEqual(INITIAL_STATE.todos);
    expect(todoItems.length).toBe(INITIAL_STATE.todos.length);
    expect(todoItems.length).toBe(wrapper.props("todos").length);
  });

  test("'todos' property should not have elements when todos are not passed to the component", async () => {
    const wrapper = await shallowMount(TodoList, {
      global: {
        mocks: {
          $route,
        },
        plugins: [store, router],
      },
      props: {
        todos: [],
      },
      computed: {
        showCheckAllTodos: () => false,
        getCompletedLength: () => false,
      },
    });
    const todoItems = await wrapper.findAll("todo-item-stub");
    expect(wrapper.props("todos")).toEqual([]);
    expect(todoItems.length).toBe(wrapper.props("todos").length);
  });

  test("double click event should call 'changeEditMode' method", async () => {
    const spyChangeEditMode = jest.spyOn(wrapper.vm, "changeEditMode");
    const todoItemList = await wrapper.findAll('[data-id-test="edit"]');
    expect(todoItemList.length > 0).toBe(true);
    await todoItemList[1].trigger("dblclick");
    expect(spyChangeEditMode).toHaveBeenCalled();
  });
});
