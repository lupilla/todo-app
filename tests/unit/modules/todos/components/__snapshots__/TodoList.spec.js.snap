// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList tests TodoList Component should match with snapshot 1`] = `
<main id="main" class="main">
  <div><input id="checkAll" type="checkbox" class="toggle-all" data-id-test="checkAll" aria-label="Check all Todos"><label for="checkAll"></label></div>
  <ul class="todo-list">
    <li data-id-test="edit" class="">
      <todo-item-stub todo="[object Object]"></todo-item-stub>
    </li>
    <li data-id-test="edit" class="">
      <todo-item-stub todo="[object Object]"></todo-item-stub>
    </li>
    <li data-id-test="edit" class="completed">
      <todo-item-stub todo="[object Object]"></todo-item-stub>
    </li>
  </ul>
</main>
`;
